src/net/network_manager.cpp

```diff
@@ -28,12 +30,38 @@ void ServerNetworkManager::connect() {
 }
 
 void ServerNetworkManager::register_listeners() {
+    // Menu phase.
+    events::menu::respond_join_event.connect([&](int conn_id, proto::JoinResponse& resp) {
+        proto::ServerMessage msg;
+        msg.set_allocated_join_response(new proto::JoinResponse(resp));
+        server.send_to(conn_id, msg);
+    });
+
     // Build phase.
     events::build::respond_build_event.connect([&](proto::Construct& c) {
         proto::ServerMessage msg;
         msg.set_allocated_build_update(new proto::Construct(c));
         server.send_to_all(msg);
     });
+
+    // Dungeon phase.
+    events::dungeon::network_positions_event.connect([&](std::map<int, Player*> & players) {
+        proto::ServerMessage msg;
+        proto::GameState* state = new proto::GameState();
+        for (auto & pair : players) {
```

Comment:
I think it would be better to keep proto construction at the dispatcher of the event, since the network manager should not have to be aware of `player_obj->transform`, etc. It makes it easier to understand if we keep this message shuffling at the call site IMO. Same for the listeners in `ClientNetworkManager::register_listeners`.

I actually think that the dispatcher should even be responsible for construction of `proto::ServerMessage`, so the manager only has to listen to one event that means "send a message to all clients," but meh.